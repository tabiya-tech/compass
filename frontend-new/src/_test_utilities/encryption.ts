import { webcrypto as crypto } from "crypto";
import { EncryptionConfig } from "src/sensitiveData/config/encryptionConfig";

// Constraining the key size to 2048 or 4096 bits for RSA as the backend expect the 4096 in calculating the maximum size of the encrypted key.
export async function generateRSACryptoPairKey(keySize: 2048 | 4096) {
  return await crypto.subtle.generateKey(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
      modulusLength: keySize,
      // 3 is a common exponent in practice
      publicExponent: new Uint8Array([1, 0, 1]),
      hash: EncryptionConfig.RSA.KEY_HASH_FN,
    },
    true, // Whether the key is extractable (i.e. can be used in exportKey)
    ["encrypt", "decrypt"] // the key can be used for both encryption and decryption
  );
}

function arrayBufferToBase64(buffer: ArrayBuffer): string {
  // @ts-ignore
  const binary = String.fromCharCode(...new Uint8Array(buffer));

  return Buffer.from(binary, "binary").toString("base64");
}

/**
 * Export a CryptoKey to a PEM formatted string (public key)
 *
 * @param key The CryptoKey to export as a PEM formatted string
 * @returns The PEM formatted string.
 */
export async function exportCryptoPublicKey(key: CryptoKey) {
  // Export the key to ArrayBuffer
  const exported = await crypto.subtle.exportKey(EncryptionConfig.RSA.FORMAT, key);

  const base64String = arrayBufferToBase64(exported);

  // add the PEM header and footer to make it a PEM formatted string.
  return `-----BEGIN PUBLIC KEY-----\n${base64String}\n-----END PUBLIC KEY-----`;
}

/**
 * Decrypt the encrypted data with the private key
 * using RSA Algorithm.
 *
 * @param privateKey
 * @param encryptedData
 */
export async function decryptWithRSA(privateKey: CryptoKey, encryptedData: ArrayBuffer) {
  return await crypto.subtle.decrypt(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
    },
    privateKey,
    encryptedData
  );
}

export async function decryptWithAES(encryptedData: ArrayBuffer, iv: ArrayBuffer, key: ArrayBuffer) {
  try {
    const importedKey = await crypto.subtle.importKey(
      EncryptionConfig.AES.FORMAT,
      key,
      EncryptionConfig.AES.ALGORITHM,
      true,
      ["decrypt"]
    );

    return await crypto.subtle.decrypt(
      {
        name: EncryptionConfig.AES.ALGORITHM,
        iv,
        length: EncryptionConfig.AES.KEY_LEN,
        tagLength: EncryptionConfig.AES.TAG_LEN,
      },
      importedKey,
      encryptedData
    );
  } catch (e) {
    console.error(e);
    throw e;
  }
}
