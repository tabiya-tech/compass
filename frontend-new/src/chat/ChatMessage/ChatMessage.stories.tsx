import type { Meta, StoryObj } from "@storybook/react";
import ChatMessage from "./ChatMessage";
import { ConversationMessageSender } from "src/chat/ChatService/ChatService.types";
import { nanoid } from "nanoid";

const meta: Meta<typeof ChatMessage> = {
  title: "Chat/ChatMessage",
  component: ChatMessage,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof ChatMessage>;

export const FromCompass: Story = {
  args: {
    chatMessage: {
      id: nanoid(),
      sender: ConversationMessageSender.COMPASS,
      sent_at: new Date().toISOString(),
      message:
        "Before we start, would you like to introduce yourself and tell me a bit about your life and what brought you here today?",
    },
  },
};

export const FromMe: Story = {
  args: {
    chatMessage: {
      id: nanoid(),
      sender: ConversationMessageSender.USER,
      sent_at: new Date().toISOString(),
      message: "Hi. I'm here to learn about my skills!",
    },
  },
};

export const LongMessage: Story = {
  args: {
    chatMessage: {
      id: nanoid(),
      sender: ConversationMessageSender.USER,
      sent_at: new Date().toISOString(),
      message:
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
    },
  },
};

export const LongUnBrokenWord: Story = {
  args: {
    chatMessage: {
      id: nanoid(),
      sender: ConversationMessageSender.USER,
      sent_at: new Date().toISOString(),
      message: "a".repeat(1000),
    },
  },
};

export const SingleLetter: Story = {
  args: {
    chatMessage: {
      id: nanoid(),
      sender: ConversationMessageSender.USER,
      sent_at: new Date().toISOString(),
      message: "a",
    },
  },
};

export const ShownWithDifferentTimestamps: Story = {
  render: () => (
    <>
      <ChatMessage
        chatMessage={{
          id: nanoid(),
          sender: ConversationMessageSender.USER,
          sent_at: new Date().toISOString(),
          message: "sent just now",
        }}
        isTyping={false}
      />
      <ChatMessage
        chatMessage={{
          id: nanoid(),
          sender: ConversationMessageSender.USER,
          sent_at: new Date(Date.now() - 1000 * 60 * 60).toISOString(),
          message: "sent an hour ago",
        }}
        isTyping={false}
      />
      <ChatMessage
        chatMessage={{
          id: nanoid(),
          sender: ConversationMessageSender.USER,
          sent_at: new Date(Date.now() - 1000 * 60 * 60 * 24).toISOString(),
          message: "sent yesterday",
        }}
        isTyping={false}
      />
      <ChatMessage
        chatMessage={{
          id: nanoid(),
          sender: ConversationMessageSender.USER,
          sent_at: new Date(Date.now() - 1000 * 60 * 60 * 24 * 2).toISOString(),
          message: "sent two days ago",
        }}
        isTyping={false}
      />
      <ChatMessage
        chatMessage={{
          id: nanoid(),
          sender: ConversationMessageSender.USER,
          sent_at: new Date(Date.now() - 1000 * 60 * 60 * 24 * 7).toISOString(),
          message: "sent a week ago",
        }}
        isTyping={false}
      />
      <ChatMessage
        chatMessage={{
          id: nanoid(),
          sender: ConversationMessageSender.USER,
          sent_at: new Date(Date.now() - 1000 * 60 * 60 * 24 * 30).toISOString(),
          message: "sent a month ago",
        }}
        isTyping={false}
      />
      <ChatMessage
        chatMessage={{
          id: nanoid(),
          sender: ConversationMessageSender.USER,
          sent_at: new Date(Date.now() - 1000 * 60 * 60 * 24 * 30 * 12).toISOString(),
          message: "sent a year ago",
        }}
        isTyping={false}
      />
    </>
  ),
};
