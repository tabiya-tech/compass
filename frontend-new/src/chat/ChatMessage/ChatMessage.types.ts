export enum ChatMessageOrigin {
  COMPASS = "compass",
  ME = "me",
}

export type ChatMessageProps = {
  origin: ChatMessageOrigin;
  message: string;
  time: Date;
};
