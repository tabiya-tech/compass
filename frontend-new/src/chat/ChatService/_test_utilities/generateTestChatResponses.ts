import { ConversationMessage, ConversationMessageSender } from "src/chat/ChatService/ChatService.types";

export const generateTestChatResponses = (): ConversationMessage[] => {
  return [
    {
      message: "Hello! How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "I can help you with that. What is your name?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
  ];
};

export const generateTestHistory = (): ConversationMessage[] => {
  return [
    {
      message: "Hello! How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "I can help you with that. What is your name?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "My name is John.",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.USER,
    },
    {
      message: "Nice to meet you, John. How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
  ];
};
