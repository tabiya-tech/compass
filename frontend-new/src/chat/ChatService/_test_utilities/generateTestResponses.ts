import { LastMessage, RootObject } from "src/chat/ChatService/ChatService.types";

export const generateRootObjectResponse = (): RootObject => ({
  last: {
    message_for_user: "Hello",
    finished: false,
    agent_type: "WelcomeAgent",
    reasoning: "Initial greeting",
    agent_response_time_in_sec: 1.0,
    llm_stats: [
      {
        error: "",
        prompt_token_count: 50,
        response_token_count: 10,
        response_time_in_sec: 1.0,
      },
    ],
  },
  conversation_context: {
    all_history: {
      turns: [
        {
          index: 1,
          input: {
            message: "Hello",
          },
          output: {
            message_for_user: "Hi there!",
            finished: false,
            agent_type: "WelcomeAgent",
            reasoning: "Greeting response",
            agent_response_time_in_sec: 0.5,
            llm_stats: [
              {
                error: "",
                prompt_token_count: 30,
                response_token_count: 10,
                response_time_in_sec: 0.5,
              },
            ],
          },
        },
      ],
    },
    history: {
      turns: [],
    },
    summary: "",
  },
});

export const generateLastMessageResponse = (): LastMessage => ({
  message_for_user: "Hi there!",
  finished: false,
  agent_type: "WelcomeAgent",
  reasoning: "Greeting response",
  agent_response_time_in_sec: 1.0,
  llm_stats: [
    {
      error: "",
      prompt_token_count: 30,
      response_token_count: 10,
      response_time_in_sec: 0.5,
    },
  ],
});
