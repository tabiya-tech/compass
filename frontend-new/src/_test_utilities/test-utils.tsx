// Based on https://testing-library.com/docs/react-testing-library/setup/
import React, { ReactElement, ReactNode } from "react";
import { render, renderHook, RenderHookOptions, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";

// Import the Firebase mock utilities
import "src/_test_utilities/firebaseMock";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/isOnlineProvider/IsOnlineProvider";
jest.mock("firebase/compat/app", () => {
  return {
    initializeApp: jest.fn(),
    auth: jest.fn().mockReturnValue({
      signOut: jest.fn(),
      currentUser: {
        getIdToken: jest.fn(),
        getTokenResult: jest.fn(),
      },
      onAuthStateChanged: jest.fn().mockReturnValue(jest.fn()),
      signInWithPopup: jest.fn(),
      GoogleAuthProvider: { PROVIDER_ID: "google.com" },
    }),
  };
});

jest.mock("firebaseui", () => {
  return {
    auth: {
      AuthUI: {
        start: jest.fn(),
        getInstance: jest.fn().mockReturnValue({
          start: jest.fn(),
          reset: jest.fn(),
        }),
        reset: jest.fn(),
      },
    },
  };
});

jest.mock("@sentry/react", () => ({
  withSentry: (Component: any) => Component,
  getFeedback: jest.fn(),
  withProfiler: (Component: any) => Component,
  wrapCreateBrowserRouter: (Component: any) => Component,
}));

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

function customRenderHook<T>(
  hook: () => T,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { wrapper?: React.ComponentType }
) {
  const CombinedWrapper = ({ children }: { children: ReactNode }) => {
    const Wrapper = options?.wrapper ?? React.Fragment;
    return (
      <AllTheProviders>
        <Wrapper>{children}</Wrapper>
      </AllTheProviders>
    );
  };

  return renderHook(hook, { wrapper: CombinedWrapper, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
export { customRenderHook as renderHook };
