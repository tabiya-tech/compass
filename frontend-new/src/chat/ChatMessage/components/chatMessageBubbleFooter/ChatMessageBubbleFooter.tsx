import React from "react";
import { Box, Divider, useTheme } from "@mui/material";
import FeedbackFormFooter from "src/feedback/feedbackForm/components/feedbackFormButton/FeedbackFormFooter";
import { ChatMessageFooterType } from "../../ChatMessage";

const uniqueId = "e4f89c2a-8c7d-4a3b-9f1c-d5c9e8f6b3a2";

export const DATA_TEST_ID = {
  CHAT_MESSAGE_BUBBLE_FOOTER: `chat-message-bubble-footer-${uniqueId}`,
  CHAT_MESSAGE_BUBBLE_FOOTER_DIVIDER: `chat-message-bubble-footer-divider-${uniqueId}`,
};

interface Props {
  footerType: ChatMessageFooterType;
  notifyOpenFeedbackForm: () => void;
}

const ChatMessageBubbleFooter: React.FC<Props> = ({ footerType, notifyOpenFeedbackForm }) => {
  const theme = useTheme();

  const getFooterFromType = (type: ChatMessageFooterType) => {
    switch (type) {
      case ChatMessageFooterType.FEEDBACK_FORM:
        return <FeedbackFormFooter notifyOpenFeedbackForm={notifyOpenFeedbackForm} />;
      default:
        return null;
    }
  };

  return (
    <Box
      data-testid={DATA_TEST_ID.CHAT_MESSAGE_BUBBLE_FOOTER}
      sx={{
        display: "flex",
        flexDirection: "column",
        gap: theme.fixedSpacing(theme.tabiyaSpacing.xs),
        width: "100%",
      }}
    >
      <Divider data-testid={DATA_TEST_ID.CHAT_MESSAGE_BUBBLE_FOOTER_DIVIDER} />
      {getFooterFromType(footerType)}
    </Box>
  );
};

export default ChatMessageBubbleFooter;
