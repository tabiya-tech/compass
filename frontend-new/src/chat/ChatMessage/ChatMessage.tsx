import React from "react";
import { Box, Typography, styled, alpha } from "@mui/material";
import { IChatMessage } from "src/chat/Chat.types";
import { getDurationFromNow } from "src/utils/getDurationFromNow/getDurationFromNow";
import { ConversationMessageSender } from "src/chat/ChatService/ChatService.types";
import ChatMessageBubbleFooter from "./components/chatMessageBubbleFooter/ChatMessageBubbleFooter";

export enum ChatMessageFooterType {
  FEEDBACK_FORM = "feedback-form",
}

const uniqueId = "2fbaf2ef-9eab-485a-bd28-b4a164e18b06";

export const DATA_TEST_ID = {
  CHAT_MESSAGE_CONTAINER: `chat-message-container-${uniqueId}`,
  CHAT_MESSAGE_TIMESTAMP: `chat-message-sent_at-${uniqueId}`,
};

const MessageContainer = styled(Box)<{ origin: ConversationMessageSender }>(({ theme, origin }) => ({
  display: "flex",
  flexDirection: "column",
  alignItems: origin === ConversationMessageSender.USER ? "flex-end" : "flex-start",
  marginBottom: theme.spacing(theme.tabiyaSpacing.sm),
  width: "100%",
}));

const MessageBubble = styled(Box)<{ origin: ConversationMessageSender }>(({ theme, origin }) => ({
  maxWidth: "80%",
  minWidth: "30%",
  variants: "outlined",
  wordWrap: "break-word",
  padding: theme.fixedSpacing(theme.tabiyaSpacing.sm),
  border: origin === ConversationMessageSender.USER ? `2px solid ${theme.palette.primary.light}` : "none",
  borderRadius: origin === ConversationMessageSender.USER ? "12px 0px 12px 12px" : "12px 12px 12px 0px",
  backgroundColor:
    origin === ConversationMessageSender.USER ? alpha(theme.palette.primary.light, 0.16) : theme.palette.grey[100],
  color: origin === ConversationMessageSender.USER ? theme.palette.primary.contrastText : theme.palette.text.primary,
  position: "relative",
  alignSelf: origin === ConversationMessageSender.USER ? "flex-end" : "flex-start",
  display: "flex",
  flexDirection: "column",
  gap: theme.fixedSpacing(theme.tabiyaSpacing.sm),
}));

const TimeStamp = styled(Typography)(({ theme }) => ({
  fontSize: theme.typography.body2.fontSize,
  color: theme.palette.text.secondary,
  marginTop: theme.spacing(theme.tabiyaSpacing.xs),
}));

type ChatMessageProps = {
  chatMessage: IChatMessage;
  notifyOpenFeedbackForm: () => void;
};

const ChatMessage: React.FC<ChatMessageProps> = ({ chatMessage, notifyOpenFeedbackForm }) => {
  let duration;
  try {
    duration = getDurationFromNow(new Date(chatMessage.sent_at));
  } catch (e) {
    console.error(new Error("Failed to get message duration", { cause: e }));
  }
  return (
    <MessageContainer origin={chatMessage.sender} data-testid={DATA_TEST_ID.CHAT_MESSAGE_CONTAINER}>
      <MessageBubble origin={chatMessage.sender}>
        <Typography whiteSpace="pre-line">{chatMessage.message}</Typography>
        {chatMessage.footerType !== undefined && (
          <ChatMessageBubbleFooter
            footerType={chatMessage.footerType}
            notifyOpenFeedbackForm={notifyOpenFeedbackForm}
          />
        )}
      </MessageBubble>
      {!chatMessage.isTypingMessage && (
        <TimeStamp data-testid={DATA_TEST_ID.CHAT_MESSAGE_TIMESTAMP} variant="caption">
          sent {duration}
        </TimeStamp>
      )}
    </MessageContainer>
  );
};

export default ChatMessage;
