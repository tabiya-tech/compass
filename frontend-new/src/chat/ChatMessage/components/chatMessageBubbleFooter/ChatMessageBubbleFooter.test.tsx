import "src/_test_utilities/consoleMock";

import { render, screen } from "src/_test_utilities/test-utils";
import ChatMessageBubbleFooter, { DATA_TEST_ID } from "./ChatMessageBubbleFooter";
import { ChatMessageFooterType } from "src/chat/ChatMessage/ChatMessage";
import { DATA_TEST_ID as FEEDBACK_FORM_BUTTON_DATA_TEST_ID } from "src/feedback/feedbackForm/components/feedbackFormButton/FeedbackFormFooter";

describe("ChatMessageBubbleFooter", () => {
  test("should render component with feedback form button", () => {
    // GIVEN a ChatMessageBubbleFooter component with feedback form button type
    const givenNotifyOpenFeedbackForm = jest.fn();

    // WHEN the component is rendered
    render(
      <ChatMessageBubbleFooter
        footerType={ChatMessageFooterType.FEEDBACK_FORM}
        notifyOpenFeedbackForm={givenNotifyOpenFeedbackForm}
      />
    );

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();

    // AND expect the footer container to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_BUBBLE_FOOTER)).toBeInTheDocument();

    // AND expect the divider to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_BUBBLE_FOOTER_DIVIDER)).toBeInTheDocument();

    // AND expect the feedback form footer to be in the document
    expect(screen.getByTestId(FEEDBACK_FORM_BUTTON_DATA_TEST_ID.FEEDBACK_FORM_FOOTER_CONTAINER)).toBeInTheDocument();

    // AND expect the component to match the snapshot
    expect(screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_BUBBLE_FOOTER)).toMatchSnapshot();
  });
});
