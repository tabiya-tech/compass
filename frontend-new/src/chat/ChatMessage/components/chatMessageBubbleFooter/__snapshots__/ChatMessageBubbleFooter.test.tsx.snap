// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessageBubbleFooter should render component with feedback form button 1`] = `
<div
  class="MuiBox-root css-1h9gn4j"
  data-testid="chat-message-bubble-footer-e4f89c2a-8c7d-4a3b-9f1c-d5c9e8f6b3a2"
>
  <hr
    class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    data-testid="chat-message-bubble-footer-divider-e4f89c2a-8c7d-4a3b-9f1c-d5c9e8f6b3a2"
  />
  <div
    class="MuiBox-root css-ckyhk3"
    data-testid="feedback-form-button-container-41675f8b-257c-4a63-9563-3fe9feb6a850"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-disableElevation MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-disableElevation css-fph7a5-MuiButtonBase-root-MuiButton-root"
      data-testid="feedback-form-button-41675f8b-257c-4a63-9563-3fe9feb6a850"
      tabindex="0"
      type="button"
    >
      Give feedback
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
